### DISCUSSION TITLE
sx1262 FSK mode, Teensy 4.0 stop working after update.

### DISCUSSION THREAD

**Discussion by Zakhidoff** [jgromes/RadioLib#1338] [General]
Hi. About 6 mounts ago successfully run example SX126x FSK Modem with some modification. Couple weeks ago after updating RadioLib library and Arduino IDE, code stop working with [SX1262] Timed out while transmitting! Any idea, why?

```c++
#include <Arduino.h>
#include <SPI.h>
#include <RadioLib.h>

// SX1262 has the following connections:
// NSS pin:   10
// DIO1 pin:  9
// NRST pin:  7
// BUSY pin:  8
SX1262 radio = new Module(10, 9, 7, 8);

int16_t val1, val2, val3, val4, val5, Srv;

struct Data { 
       int16_t var[16];
       };

void setup() {
  Serial.begin(115200);
  
  int state = radio.beginFSK();
  if (state == RADIOLIB_ERR_NONE) { Serial.println(F("success!")); } 
  else 
     { Serial.print(F("failed, code "));
       Serial.println(state);
       while (true); }
   
  radio.setFrequency(915.0);
  radio.setBitRate(300.0);
  radio.setFrequencyDeviation(50.0);
  radio.setRxBandwidth(467.0);
  radio.setOutputPower(22.0);
  radio.setCurrentLimit(100.0);
  //radio.setDataShaping(RADIOLIB_SHAPING_1_0);
  //radio.setDataShaping(RADIOLIB_SHAPING_NONE);
  //uint8_t syncWord[] = {0x01, 0x23, 0x45, 0x67,0x89, 0xAB, 0xCD, 0xEF};
  //radio.setSyncWord(syncWord, 8);
  radio.setRfSwitchPins(15, 14);
  //radio.setPreambleLength(16);
  //radio.setNodeAddress(0x23);
  //radio.setBroadcastAddress(0x45);
  //radio.disableAddressFiltering();
  //radio.setSyncBits(syncWord, 8);
    radio.setCRC(0, 0xFFFF, 0x8005, true);
  
  if (state != RADIOLIB_ERR_NONE) {
    Serial.print(F("Unable to set configuration, code "));
    Serial.println(state);
    while (true); }
}

int16_t count = 0;

void loop() {
  val1  = analogRead(A9);
  val2  = analogRead(A8);
  val3  = analogRead(A7);
  val4  = analogRead(A6);
  val5  = analogRead(A5);

  //Srv = map (val2, 0, 1023, 1000,2000);
  Srv = map (val2, 0, 1023, 0,180);

  Data datasend = {val2, val1, count++ ,Srv, 1900, val3,3000, val4, 5000, val5};
  int state = radio.transmit((byte*)&datasend, sizeof(datasend));   

    if (state == RADIOLIB_ERR_NONE) 
            { Serial.println(F("[SX1262] Packet transmitted successfully!"));}
    else if (state == RADIOLIB_ERR_PACKET_TOO_LONG) 
            { Serial.println(F("[SX1262] Packet too long!")); } 
    else if (state == RADIOLIB_ERR_TX_TIMEOUT)
            { Serial.println(F("[SX1262] Timed out while transmitting!"));} 
    else    { Serial.println(F("[SX1262] Failed to transmit packet, code "));
              Serial.println(state); }

    // print measured data rate
    Serial.print(F("[SX1262] Datarate:\t"));
    Serial.print(radio.getDataRate());
    Serial.println(F(" bps"));
 }
```

**Comment by radiohound** [comment]
It looks like your setCurrentLimit is set below the specified power that it takes for that device to transmit at 22db. You could try raising the setCurrentLimit or removing it altogether. Another thing to try is [debug mode](https://github.com/jgromes/RadioLib/wiki/Debug-mode) .

**Comment by jgromes** [comment]
@Zakhidoff 

> Couple weeks ago after updating RadioLib library and Arduino IDE, code stop working with [SX1262] Timed out while transmitting! Any idea, why?

Not really, since we have no information about the version before and after the update, nor the debug output as @radiohound pointed out. The first thing I would try is to go back to the original version. If it still doesn't work, a hardware problem is likely. After that, all the classic debugging steps - try to get to the simplest code that still manifests the issue.

**Reply by Zakhidoff** [reply]
hi. Thanks for quick respond. We had 2 sets of working prototypes. after some updating first transmitter stop working. im also upload original code without any updates, same result. In Lora mode everything works fine, but in FSK mode [SX1262] Timed out while transmitting! 

```c++
#include <Arduino.h>
#include <SPI.h>
#include <RadioLib.h>

// SX1262 has the following connections:

// NSS pin:   10
// DIO1 pin:  9
// NRST pin:  7
// BUSY pin:  8
SX1262 radio = new Module(10, 9, 7, 8);

int16_t val1, val2, val3, val4, val5;

struct Data { int16_t var[16]; } ;
int16_t count = 0; // counter to keep track of transmitted packets

 void setup() {
  Serial.begin(115200);

  // initialize SX1262 with default settings
  Serial.print(F("[SX1262] Initializing ... "));
  int state = radio.begin();
  if (state == RADIOLIB_ERR_NONE) { Serial.println(F("success!")); } 
  else 
     { Serial.print(F("failed, code "));
       Serial.println(state);
       while (true); }

  radio.setFrequency(915.5);
  radio.setBandwidth(500.0);
  radio.setSpreadingFactor(6);
  radio.setCodingRate(5);
  radio.setSyncWord(0x34);
  radio.setOutputPower(22);
  radio.setPreambleLength(8);
  radio.setCRC(false);
  radio.setRfSwitchPins(15, 14);   //RX enable: 15, TX enable: 14;

//int state = radio.begin(915.0, 500.0, 6, 1, 0x34, 22, 8); 
}

void loop() {
  
  val1  = analogRead(A9);
  val2  = analogRead(A8);
  val3  = analogRead(A7);
  val4  = analogRead(A6);
  val5  = analogRead(A5);
  
  Serial.print(F("[SX1262] Transmitting packet ... "));
 
    Data datasend = {1500,val1 ,count++,val2,1900, val3,3000, val4, 5000, val5}; 
    int state = radio.transmit((byte*)&datasend, sizeof(datasend));
    
    if (state == RADIOLIB_ERR_NONE) {
    // the packet was successfully transmitted
    Serial.println(F("success!"));

    // print measured data rate
    Serial.print(F("[SX1262] Datarate:\t"));
    Serial.print(radio.getDataRate());
    Serial.println(F(" bps"));

  } 
```

**Reply by jgromes** [reply]
@Zakhidoff I don't understand what was updated - RadioLib, Arduino IDE, or your code? Or all of them? If you did change RadioLib version, what was the original version and what was the udpated one?

I also note that you have neither provided the debug output, nor have tried to simplify the sketch that is actually causing the issue. None of us have your hardware, so if you don't do these things, we can't help you.

EDIT: I have also updated your last two posts, to actually make the code readable. Please take a look at the edit I did, it's very simple to format code here.

**Reply by Zakhidoff** [reply]
Hi. Thanks for respond. I did update  Arduino IDE ( from 1.8.3 to 2.3.3) . I don't know exactly library version number but last working code was on April 2024. Hardware that I'm use, its Teensy 4.0 and Waveshare Core1262 HF LoRa Module, SX1262 Chip. Also tried E22 900M30S. Same result. In LORA mode both module works fine, in FSK mode error -5.   Debug output??? 

23:23:09.448 -> [SX1262] Timed out while transmitting!
23:23:09.448 -> [SX1262] Datarate:	0.00 bps
23:23:09.448 -> RLB_SPI: CMDW	80	
23:23:09.448 -> RLB_SPI: SI		0	
23:23:09.448 -> RLB_SPI: SO	A2	A2	
23:23:09.448 -> RLB_SPI: CMDR	11	
23:23:09.448 -> RLB_SPI: SI		0	0	
23:23:09.448 -> RLB_SPI: SO	A2	A2	0	
23:23:09.448 -> RLB_DBG: Timeout in 12 ms
23:23:09.448 -> RLB_SPI: CMDR	11	
23:23:09.448 -> RLB_SPI: SI		0	0	
23:23:09.448 -> RLB_SPI: SO	A2	A2	0	
23:23:09.448 -> RLB_SPI: CMDW	8C	
23:23:09.448 -> RLB_SPI: SI		0	10	1	0	0	20	FF	10	1	
23:23:09.448 -> RLB_SPI: SO	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	
23:23:09.448 -> RLB_SPI: CMDW	8	
23:23:09.448 -> RLB_SPI: SI		2	1	0	1	0	0	0	0	
23:23:09.448 -> RLB_SPI: SO	A2	A2	A2	A2	A2	A2	A2	A2	A2	
23:23:09.448 -> RLB_SPI: CMDW	8F	
23:23:09.448 -> RLB_SPI: SI		0	0	
23:23:09.448 -> RLB_SPI: SO	A2	A2	A2	
23:23:09.448 -> RLB_SPI: CMDW	E	0	
23:23:09.448 -> RLB_SPI: SI			3	0	B7	1	52	7	1	0	6C	7	EE	0	B8	B	98	1	88	13	A2	1	0	0	0	0	0	0	0	0	0	0	0	0	
23:23:09.448 -> RLB_SPI: SO	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	A2	
23:23:09.448 -> RLB_SPI: CMDW	2	
23:23:09.448 -> RLB_SPI: SI		43	FF	
23:23:09.448 -> RLB_SPI: SO	A2	A2	A2	
23:23:09.448 -> RLB_SPI: CMDR	1D	8	89	
23:23:09.448 -> RLB_SPI: SI				0	0	
23:23:09.448 -> RLB_SPI: SO	A2	A2	A2	A2	4

**Reply by jgromes** [reply]
> last working code was on April 2024

That does not help me at all, how should I know which version of the library you were using at theat point?

The debug log is inconsistent with what you posted above. In your FSK code, you are sending what looks like 20 bytes of data at 300 kbps, but the debug log shows 12 ms timeout, which doesn't add up.

I tried replicating the issue using ESP32 with SX1262. For very fast data rates such as 300 kbps, the calculated timeout is very short (1 ms), which did cause timing issues on the ESP32 since the timer used to measure the timeout has a millisecond precision. I changed the timeout calculation to make sure the timeout never drops below 5 ms which should resolve this. If the problem persists, try interrupt-driven transmission and if that doesn't work either, it is most likely bad wiring since the end of transmission is signalled via the DIO1 pin.

**Reply by Zakhidoff** [reply]
Hi. Thanks again for respond. Let update my question. I have successfully working setup for teensy 4.1 and sx1262 in LORA mode and doesn't working in FSK mode. I'm use library example with updated pin connection. I guess, maybe somethings wrong with spi speed or timeout duration. If i had any bad wiring DIO1 pin,  LORA mode also doesn't work. Probably library in FSK mode doesn't correctly work with Teensy boards. i didn't try sx1262 module with another boards, maybe it will work.

**Reply by jgromes** [reply]
@Zakhidoff FYI we have found and fixed an issue in SX126x FSK mode (see #1350) taht could be related to this issue, so now is the right time to udpate your library to the master and see if the issue still persists.
